pkg/arvo/lib/eth-provider.hoon

```diff
@@ -0,0 +1,187 @@
+  (pure:m -)
+++  get-tx-by-hash
+  |=  tx-hash=@ux
+  =/  m  (strand:strandio ,transaction-result:rpc:ethereum)
+  ;<  out=ethout:eth-provider  bind:m
+  (select-provider-mode [%get-tx-by-hash tx-hash])
+  ?>  ?=(%get-tx-by-hash -.out)
+  =+  `transaction-result:rpc:ethereum`+.out
+  (pure:m -)
+++  get-logs-by-hash
+  |=  [=hash:block contracts=(list address) =topics:eth-provider]
+  =/  m  (strand:strandio ,(list event-log:rpc:ethereum))
+  ;<  out=ethout:eth-provider  bind:m
+  (select-provider-mode [%get-logs-by-hash hash contracts topics])
+  ?>  ?=(%get-logs-by-hash -.out)
+  =+  `(list event-log:rpc:ethereum)`+.out
+  (pure:m -)
+++  get-logs-by-range
+  |=  $:  
+          contracts=(list address)
+          =topics:eth-provider
+          from=number:block
+          to=number:block
+      ==
+  =/  m  (strand:strandio ,(list event-log:rpc:ethereum))
+  ;<  out=ethout:eth-provider  bind:m
+  (select-provider-mode [%get-logs-by-range contracts topics from to])
+  ?>  ?=(%get-logs-by-range -.out)
+  =+  `(list event-log:rpc:ethereum)`+.out
+  (pure:m -)
+++  get-next-nonce
+  |=  =address
+  =/  m  (strand:strandio ,@ud)
+  ;<  out=ethout:eth-provider  bind:m
+  (select-provider-mode [%get-next-nonce address])
+  ?>  ?=(%get-next-nonce -.out)
+  =+  `@ud`+.out
+  (pure:m -)
+++  get-balance
+  |=  =address
+  =/  m  (strand:strandio ,@ud)
+  ;<  out=ethout:eth-provider  bind:m
+  (select-provider-mode [%get-balance address])
+  ?>  ?=(%get-balance -.out)
+  =+  `@ud`+.out
+  (pure:m -)
+::
+++  select-provider-mode
+  =,  strand=strand:spider
+  =,  dejs-soft:format
+  |=  eth-input=ethin:eth-provider
+  =/  m  (strand ,ethout:eth-provider)
+  ^-  form:m
+  ;<  =bowl:spider  bind:m  get-bowl:strandio
+  ?>  =(src.bowl our.bowl)
+  ;<    state=state:eth-provider
+      bind:m
+    (scry:strandio state:eth-provider /gx/eth-provider/get-state/noun)
+  ?-  active.state
+    %local
+  (call-ethio eth-input active.state url.local.state)
+    %provider
+  (call-ethio eth-input active.state url.provider.state)
+    %client
+  ::  TODO subscribe to our thread id
+  ;<  ~  bind:m  
+    (watch:strandio [%updates tid.bowl ~] [provider.client.state %eth-provider] [%updates tid.bowl ~])
+  ;<  ~  bind:m
+    %+  poke:strandio
+      [provider.client.state %eth-provider] 
+    [%provider-action !>([%provide tid.bowl eth-input])]
+  ;<  =cage  bind:m  (take-fact:strandio [%updates tid.bowl ~])
+  (pure:m !<(ethout:eth-provider q.cage))
+  ==
+++  call-ethio
+  |=  [arg=ethin:eth-provider active=active:eth-provider url=@ta]
+  =,  strand=strand:spider
+  =/  m  (strand ,ethout:eth-provider)
+  ;<  =bowl:spider  bind:m  get-bowl:strandio
+  ?-  -.arg
+    %request-rpc
+  ;<  out=json  bind:m  (request-rpc:ethio url +.arg)
+  (pure:m [%request-rpc out])
+    %request-batch-rpc-strict
+  ;<  out=(list [id=@t =json])  bind:m
+    (request-batch-rpc-strict:ethio url +.arg)
+  (pure:m [%request-batch-rpc-strict out])
+    %request-batch-rpc-loose
```

Comment:
Big `?-` blocks like these would be easier to read with better indentation and some spacing:

```suggestion
  ?-  -.arg
      %request-rpc
    ;<  out=json  bind:m  (request-rpc:ethio url +.arg)
    (pure:m [%request-rpc out])
  ::
      %request-batch-rpc-strict
    ;<  out=(list [id=@t =json])  bind:m
      (request-batch-rpc-strict:ethio url +.arg)
    (pure:m [%request-batch-rpc-strict out])
  ::
      %request-batch-rpc-loose
    etc...
```